src/discordcr/dca.cr

```diff
@@ -0,0 +1,135 @@
+require "json"
+
+module Discord
+  # Parser for the DCA file format, a simple wrapper around Opus made
+  # specifically for Discord bots.
+  class DCAParser
+    # Magic string that identifies a DCA1 file
+    DCA1_MAGIC = "DCA1"
+
+    # The parsed metadata, or nil if it could not be parsed.
+    getter metadata : DCA1Mappings::Metadata?
+
+    # Create a new parser. It will read from the given *io*. If *raw* is set,
+    # the file is assumed to be a DCA0 file, without any metadata. If the file's
+    # metadata doesn't conform to the DCA1 specification and *strict_metadata*
+    # is set, then the parsing will fail with an error; if it is not set then
+    # the metadata will silently be `nil`.
+    def initialize(@io : IO, raw = false, @strict_metadata = false)
+      unless raw
+        verify_magic
+        parse_metadata
+      end
+    end
+
+    # Reads the next frame from the IO. If there is nothing left to read, it
+    # will return `nil`.
+    def next_frame : Bytes?
+      begin
+        header = @io.read_bytes(Int16, IO::ByteFormat::LittleEndian)
+        raise "Negative frame header (#{header} < 0)" if header < 0
+
+        buf = Bytes.new(header)
```

Comment:
The size of this buffer is restricted by `Int16::Max` (and it's small), so it would be pretty sensible I think to allocate this buffer once and reuse it. However this would mean the slice returned by this method would be only valid until the next call. Perhaps an option `reuse_buffer = false` for this method and lazy allocate the buffer?